### PR TITLE
Fix for WFCORE-4923 read-resource operation does not resolve expressi…

### PR DESCRIPTION
**Pull request by swetapatra**
…ons recursively, https://issues.redhat.com/browse/WFCORE-4923

Thanks for submitting your Pull Request!

Please delete this text, and add a link to the Jira issue solved by this PR.

Remember to use the Jira issue ID in the PR title, and any commits.
@jmesnil 
@darranl ,could you please check this now.

### CODE REVIEW

**Comment by bstansberry** [comment]
Hi @swetapatra,

I think a test should be added to org.jboss.as.controller.test.GlobalOperationsTestCase to cover this. AbstractGlobalOperationsTestCase.initModel is where the model structure those tests use is set up. One or more new resources would be needed. A test can set system properties to provide resolvable values before invoking the read-resource operation.

**Comment by swetapatra** [comment]
> Hi @swetapatra,
> 
> I think a test should be added to org.jboss.as.controller.test.GlobalOperationsTestCase to cover this. AbstractGlobalOperationsTestCase.initModel is where the model structure those tests use is set up. One or more new resources would be needed. A test can set system properties to provide resolvable values before invoking the read-resource operation.

@bstansberry thanks for reviewing my request ,I have updated the pull request with another commit containing changes to two test classes,please review and let me know if anything else is required.

**Comment by jmesnil** [comment]
@swetapatra that's not really what Brian suggested. 

You should add a test that read a resource recursively with resolve-expressions set to `true`.

You could for example duplicate org.jboss.as.controller.test.GlobalOperationsTestCase#testRecursiveReadSubModelOperationComplex and read-resource from `ModelNode operation = createOperation(READ_RESOURCE_OPERATION, "profile", "profileA");` with `recursive` and `resolve-expressions` set to `true`.

In this set set the System property `expr` to something and then check that the value of the attribute `expression` of the resource `/profile=profileA/subsystem=subsystem2` has been properly resolved with the value specified in the System property `expr`

**Comment by swetapatra** [comment]
> @swetapatra that's not really what Brian suggested.
> 
> You should add a test that read a resource recursively with resolve-expressions set to `true`.
> 
> You could for example duplicate org.jboss.as.controller.test.GlobalOperationsTestCase#testRecursiveReadSubModelOperationComplex and read-resource from `ModelNode operation = createOperation(READ_RESOURCE_OPERATION, "profile", "profileA");` with `recursive` and `resolve-expressions` set to `true`.
> 
> In this set set the System property `expr` to something and then check that the value of the attribute `expression` of the resource `/profile=profileA/subsystem=subsystem2` has been properly resolved with the value specified in the System property `expr`

I understand it now,I thought I just need to add it as a param before the test starts that way it will be available,thanks for clarification,I am working on it now

File: controller/src/test/java/org/jboss/as/controller/test/GlobalOperationsTestCase.java [hunk 1]
```diff
@@ -954,6 +954,17 @@ public void testReadResourceAttributesOnly() throws Exception {
         executeForFailure(operation);
     }
 
+    @Test
```
Comment (jmesnil) [inline controller/src/test/java/org/jboss/as/controller/test/GlobalOperationsTestCase.java:957]: that's not what I suggested. your test is creating a resource with resolve-expressions set to `true`. This has no effect.

What you should do instead is something like:


```
    @Test
    public void testRecursiveReadSubModeWithResolveExpression() throws Exception {
        try {
            // TODO: set the system property expr to a random value
            String expr = ....
            System.setProperty("expr", expr);

            ModelNode operation = createOperation(READ_RESOURCE_OPERATION, "profile", "profileA");
            operation.get(RECURSIVE).set(true);
            operation.get(RESOLVE_EXPRESSIONS).set(true);

            ModelNode result = executeForResult(operation);
            assertNotNull(result);

           // check that the sub resource subsystem=subsystem2 has an attribute expression whose value correspond to the exp system property
           assertEquals(expr, result.get("subsystem", "subsystem2", "expression").asString());     
          } finally {
               // TODO unset the system property expr 
    }
```

Comment (swetapatra) [reply controller/src/test/java/org/jboss/as/controller/test/GlobalOperationsTestCase.java:957]: @jmesnil Thank you for reviewing it,I will take a look and fix the further quoted issues

**Comment by darranl** [comment]
Hi @swetapatra how is the update to this PR going, are you able to submit the requested changes so we can get this merged and out of the queue?

**Comment by swetapatra** [comment]
> Hi @swetapatra how is the update to this PR going, are you able to submit the requested changes so we can get this merged and out of the queue?

I did get stuck on this,let me check it again and update the PR,I will get it done shortly

**Comment by swetapatra** [comment]
> > Hi @swetapatra how is the update to this PR going, are you able to submit the requested changes so we can get this merged and out of the queue?
> 
> I did get stuck on this,let me check it again and update the PR,I will get it done shortly

I have updated the test case,kindly check and let me know if still some changes are required.

**Comment by swetapatra** [comment]
@jmensil,@darranl  I have updated the pull request but there are few integration errors,should I rebase and try again?

**Comment by darranl** [comment]
I would say it does make sense to rebase at this point as this has been open for a while so that will get you up to date with master.

I am not a big fan of asking for commits to be squashed but if you are rebasing anyway it may be worth considering if the commits can be squashed into a smaller logical set of commits.

**Comment by swetapatra** [comment]
Hi @jmensil

I tried rebasing but there were too many commits and it was getting difficult to structure those.
Thus created new pull request :https://github.com/wildfly/wildfly-core/pull/4297 by taking pull from current master containing all changes.
Please review it and let me know if any further changes are required

**Comment by swetapatra** [comment]
Hi @jmensil

I tried rebasing but there were too many commits and it was getting difficult to structure those.
Thus created new pull request :https://github.com/wildfly/wildfly-core/pull/4297 by taking pull from current master containing all changes.
Please review it and let me know if any further changes are required

**Comment by swetapatra** [comment]
Hi @jmensil

I tried rebasing but there were too many commits and it was getting difficult to structure those.
Thus created new pull request :https://github.com/wildfly/wildfly-core/pull/4297 by taking pull from current master containing all changes.
Please review it and let me know if any further changes are required

**Comment by darranl** [comment]
So if rebasing is problematic that may also be made easier by squashing your commits down into either one commit for the fix and one commit for the test or squash it down to a single commit for the complete PR.

A command such as "git pull --rebase upstream master" (Assuming the remote of WildFly is called "upstream") would first remove your commits from your topic branch - apply everything that as been applied to master and then apply your commits again - any conflict should just be in those final one or two commits of yours.

When you say there are too many commits I wonder if maybe your rebase has happened the wrong way round?

**Comment by bstansberry** [comment]
This has been superseded by #4297

**Pull request closed without merging**